### PR TITLE
Add docker

### PR DESCRIPTION
**Pull request by LachlanCourt**

### CODE REVIEW

**Comment by LachlanCourt** [comment]
I'll be removing the build and push test for PRs but I have it here to test and bugfix before merging into main

File: .github/workflows/build-and-push-main.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+jobs:
+  build-and-push:
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v3
+    - name: Build Docker Image
+      run: |
+        docker build -t rainbowbot:build .
+    - name: Login to GitHub Container Registry
```
Comment (Y0sh1dk) [inline .github/workflows/build-and-push-main.yml:20]: Logging into the container registry should be the first step in the pipeline, uncase the login fails we dont waste any time building the image unnecessarily

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,16 @@
+FROM python:3.12-rc-alpine as base
```
Comment (Y0sh1dk) [inline Dockerfile:1]:
```suggestion
ARG PYTHON_VERSION=3.12

FROM python:${PYTHON_VERSION}-rc-alpine as base
```

Due to the python version being referenced in multiple places at build time, it should be an `ARG`. `${PYTHON_VERSION}` can also be used further down

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,10 @@
+build:
```
Comment (Y0sh1dk) [inline Makefile:1]: This Makefile should also accept env vars so it can be used in the pipeline

**Review by Y0sh1dk — CHANGES_REQUESTED** [review]

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+	docker build -t rainbowbot:generic .
+run:
+	docker run --env-file .env rainbowbot:generic python3 bot.py
+clean:
+	docker container rm -f $(shell docker container ls -aq)
+push:
```
[the remark this answers is not in the document]
Comment (LachlanCourt) [reply Makefile:7]: Because I needed a generic tag? 😅
Mostly borrowed from NUbots - is there a better one semantically?

**Review by Y0sh1dk — DISMISSED** [review]
LGTM

**Review by Y0sh1dk — DISMISSED** [review]

**Pull request merged**